### PR TITLE
Modernizing build infrastructure and switching to GitHub Actions

### PR DESCRIPTION
**Pull request by dfm**
A work in progress

### CODE REVIEW

**Comment by henrysky** [comment]
Hi @dfm,

I have just tried this branch. Really thanks for working on it as I need this package for my extragalactic grad course and don't want to use Linux. Anyway one small request is can you do `import setuptools._distutils` before `from numpy.distutils.core import setup, Extension` so that people can do `python setup.py develop` and do `pip uninstall fsps` properly as the ancient `numpy.distutils` does not play well with `pip uninstall`

**Comment by dfm** [comment]
@henrysky will do! Hopefully we'll have a new release soon.

**Comment by dfm** [comment]
@henrysky: Perhaps you can take a look at this branch and see if you can understand why the Windows build isn't working. I've made a bunch of changes to try to make packaging better and mac and linux have been no problem, but windows is driving me nuts! I've tried about a bunch of different settings: MinGW vs MSVC, conda vs not, etc. and I seem to always get a:

```
ImportError: DLL load failed while importing _fsps: The specified module could not be found.
```

when importing. I don't have much familiarity with Windows and I'm trying to debug using Actions, so perhaps you have a better sense?

**Comment by henrysky** [comment]
@dfm I have tried to trace what the `.pyd` trying to load and found out it is trying to load a dll which is located under `\python-fsps\fsps\.libs`. Copying the `.dll` file under `\python-fsps\fsps\.libs` to `python-fsps\src\fsps` everything runs fine.

A side note, I initially get an error from this, but I just deleted that line to get it running

``
c:\users\henry\documents\github\python-fsps\src\fsps\__init__.py in <module>
     28 os.environ["SPS_HOME"] = os.path.join(here, "libfsps")
     29
---> 30 from .fsps_version import version as __version__  # noqa
     31 from .fsps import StellarPopulation  # noqa
     32 from .filters import find_filter, get_filter, list_filters  # noqa

ModuleNotFoundError: No module named 'fsps.fsps_version'
``

**Comment by dfm** [comment]
Thanks - I'd found that too, and this is supposed to fix that: https://github.com/dfm/python-fsps/blob/b92c1765fa9502be75e889aca3af9f5119d8f95f/src/fsps/__init__.py#L16

Now, what happens when you run the tests? I have the install an import working on my end, but pytest seems to crash silently when trying to execute the tests.

(Re: the version issue, you should install using `pip install .` or `pip install -e .` instead of setup.py directly. I think you could install `setuptools_scm` manually if you prefer.)

**Comment by henrysky** [comment]
I dont think that fix is working on my end and the `.dll` is still at the wrong place. I currently do not have any idea/fix. Anyway I copied the `.dll` to where it is supposed to be and ran the tests and all passed.

The tests actually ran quite long on my 6 cores i7 laptop...Not sure if it is expected or just poor performance of gfortran on windows
image

**Comment by dfm** [comment]
Thanks @henrysky! That's very helpful. Yes - the tests are very slow :D 

Just to confirm: that hack doesn't move the `.dll`, it updates the library search path. I found that it does work when installed but not in editable mode. How did you install in the end? `python setup.py develop`?

**Comment by dfm** [comment]
OK, the scope definitely drifted on this a little bit so this is a pretty major PR in the end. Here's the summary:

1. The CI infrastructure is switched to GitHub actions instead of Travis since Travis is pretty much dead. As part of this, we're now testing on Ubuntu, Mac, and Windows.

2. I also added FSPS as a submodule and defer to using numpy's distutils tools rather than the pile o' hacks that we had for compiling before. One nice thing is that this now supports `pip install` without first editing Makefiles and futzing with all that. This is nice because the Python installation needs to know how to compile Fortran anyways, so we might as well just reuse that logic! This also has the added benefit that we can build wheels, but that'll come in a future PR. One thing that this *doesn't* change is that we'll still require users to have an FSPS clone for the data files. It would be possible to package the data files alongside this package, but that leads to a pretty big file and it seemed good to leave folks the option of having the data files stored somewhere else.

3. It turned out that the way numpy compiles Fortran libraries doesn't play super nice with Windows (see the back and forth with @henrysky above) but we got it working! I've tested using MSVC and MinGW compilers, and both installed and editable packages (I do have a Windows machine, but I'd never used it for development, but I learned :D). All seem to work now, although the code is somewhat slower on Windows for reasons that I don't totally understand. I'm satisfied with how it works now, and I'm sick of adding more hacks to the `setup.py` (most of this is for Windows specifically, Mac and Linux work out of the box...).

@bd-j: Can you take a look at this when you have some time? If you're happy with it, I'll squash and merge and then start working on drafting a new release. If not, I'm happy to make changes!

**Comment by dfm** [comment]
(PS I love it when a "quick hack" turns into a full 1.5 days of work :D)

**Comment by bd-j** [comment]
@dfm Thank you so much for this! This sounds awesome. I'll give it a closer look ASAP.

File: docs/installation.rst [hunk 1]
```diff
@@ -4,85 +4,88 @@ Installation
...
 
-Python-FSPS is being actively developed on GitHub so it's usually best
-to use the most recent development version.
-You can do this by cloning the `python-fsps repository
-<https://github.com/dfm/python-fsps>`_ and building::
+Python-FSPS is being actively developed on GitHub so you can got there to get
+the most recent development version. You can do this by cloning the `python-fsps
+repository <https://github.com/dfm/python-fsps>`_ and building:
+
+..code-block:: bash
```
Comment (bd-j) [inline docs/installation.rst:43]: missing a space after the ..

**Comment by bd-j** [comment]
Hi @dfm, I had a look, installed, and used it in other projects, seems great. This looks like a much cleaner way to integrate the fsps code for users, thanks so much!

I'll have to read up on submodules - I do some futzing about with both the fsps src and data files on occasion. One thing that I think other users may want to do is switch the isochrone and stellar libraries, which (unfortunately) requires editing pre-compiler flags and recompiling fsps.  I tested this works by changing the source in  `src/fsps/libfsps/src/` and reinstalling with `python pip -m .`, but I think this requires the development install (not PyPI)?  I could add some words about this in the install docs.  Is it possible, in principle, to give a flag to pip that causes the relevant changes to be made to the fsps submodule source when installing with pypi?

And thanks for all the fixed typos and formatting ;)

**Comment by bd-j** [comment]
Also it looks like git is no longer called anywhere, since the exact fsps hash is now handled by the submodule, so we can remove some of that text from the installation troubleshooting section (and I think this solves #115 as well)

**Comment by dfm** [comment]
@bd-j: Thanks!

>  Is it possible, in principle, to give a flag to pip that causes the relevant changes to be made to the fsps submodule source when installing with pypi?

I think that it's probably possible to pass compiler flags from the command line when doing `pip install`. Can you give me an example and I'll add a test? Depending on how common that use case is, I expect that it's still worth switching to this mode, but you definitely will have a better sense of that! Let me play around with it and then I can draft some text for the installation docs and get your feedback.

> Also it looks like git is no longer called anywhere, since the exact fsps hash is now handled by the submodule, so we can remove some of that text from the installation troubleshooting section (and I think this solves #115 as well)

Good point. I just removed the troubleshooting section for now. I expect that there could still be an issue where the data in the `SPS_HOME` directory is out of sync with the compiled code, but I thought that that was low enough risk that it was better to just skip that check. I did add a check for the only data file that I could find that was referenced in the commits that we were previously checking: https://github.com/dfm/python-fsps/blob/75e2792ee04027770a38d08a1ce706146b325ecb/src/fsps/sps_home.py#L16

**Comment by bd-j** [comment]
@dfm I agree it's still worth switching to this mode, regardless of whether pip install flags end up working; just think adding some doc text will make responding to likely issues easier. Also agree the data out-of-sync issue is low risk.

I'm in favor of merging this PR now, and dealing with the docs and any pip install flags subsequently.

Re those flags, this might be best in a separate issue, but I think it'd be something like
```
pip install fsps --isoc PADOVA --speclib BASEL
python -c "import fsps; sps = fsps.StellarPopulation(); assert sps.libraries[0] == b'pdva'; assert sps.libraries[1] == b'basel'"
```
But then somehow during the pip install the `src/fsps/libfsps/src/sps_vars.f90` lines 8 to ~ 50 would have to be edited before compiling, due to the kind of clunky way these changes are handled in the fortran.  Again, this is not high priority, especially if we have some warnings/instructions in the install docs about how to do this with the dev install.  Just wondered if it was easy or painful via pip flags.

**Comment by dfm** [comment]
Thanks - to get that behavior, we would use something like this for a local version:

```bash
FFLAGS="-DMIST=0 -DPADOVA=1 -DMILES=0 -DBASEL=1" python -m pip install .
```

And I *think* this will also work with a version from PyPI using the following:

```bash
FFLAGS="-DMIST=0 -DPADOVA=1 -DMILES=0 -DBASEL=1" python -m pip install fsps --no-binary fsps
```

Where the `--no-binary fsps` flag is required to force building from source.

**Comment by dfm** [comment]
I've added a test for the former so I think I'm happy for this to be merged. I'll do that and then open separate issues for docs about compiler flags and for drafting a release.

**Pull request merged**